src/Messenger/Transport/DsnParser.php

```diff
@@ -4,28 +4,131 @@
+        }
+
+        // check for extra keys in options
+        $optionsExtraKeys = array_diff(array_keys($options), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < count($optionsExtraKeys)) {
+            throw new InvalidArgumentException(sprintf(
+                'Unknown option found: [%s]. Allowed options are [%s].',
+                implode(', ', $optionsExtraKeys),
+                implode(', ', array_keys(self::DEFAULT_OPTIONS))
+            ));
+        }
+
+        // check for extra keys in query
+        $queryExtraKeys = array_diff(array_keys($query), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < count($queryExtraKeys)) {
+            throw new InvalidArgumentException(sprintf(
+                'Unknown option found in DSN: [%s]. Allowed options are [%s].',
+                implode(', ', $queryExtraKeys),
+                implode(', ', array_keys(self::DEFAULT_OPTIONS))
+            ));
+        }
+
+        $options = $query + $options + self::DEFAULT_OPTIONS;
+        /** @var array{
+         *      shared_access_key_name: string|null,
+         *      shared_access_key: string|null,
+         *      namespace: string|null,
+         *      entity_path: string|null,
+         *      subscription: string|null,
+         *      token_expiry: int|string,
+         *      receive_mode: AzureTransport::RECEIVE_MODE_*,
+         * } $options
+         */
+
+        $options['shared_access_key_name'] = $this->pickOption('shared_access_key_name', 'user', $parsedUrl, $options);
+        $options['shared_access_key'] = $this->pickOption('shared_access_key', 'pass', $parsedUrl, $options);
+        $options['namespace'] = $this->pickOption('namespace', 'host', $parsedUrl, $options);
+        $options['token_expiry'] = (int) $options['token_expiry'];
+
+        // Missing topic or queue name
+        if (null === $options['entity_path']) {
+            throw new InvalidArgumentException(
+                sprintf('Missing entity_path (queue or topic) for the "%s" transport.', $transportName),
+                1643989596
+            );
+        }
+
+        // Invalid receive mode
+        if (false === in_array($options['receive_mode'], self::RECEIVE_MODES, true)) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'Invalid "%s" receive_mode for the "%s" transport. It must be one of: %s.',
+                    $options['receive_mode'],
+                    $transportName,
+                    implode(', ', self::RECEIVE_MODES)
+                ),
+                1643994036
+            );
+        }
+
+        // @phpstan-ignore-next-line after all those array merges PHPstan gets confused about the result shape
+        return $options;
+    }
+
+    /**
+     * @param array<string, int|string> $dsnParts
+     * @param array<string, mixed> $options
+     * @return mixed
+     */
+    private function pickOption(string $optionName, string $dsnName, array $dsnParts, array $options)
```

Comment:
I saw that this method was only used 3 times for options that **MUST** be defined and are of string type. I'd change the return type to `string` , cast to string on line 129 and replace the line 132 with an exception:
```php
if (isset($options[$optionName])) {
    return strval($options[$optionName]);
}

throw new InvalidArgumentException(
    sprintf('Missing %s for the "%s" transport.', $optionName, $transportName),
    1702724695
);
```
The exception is the reason I added `$transportName`. I don't think that line 132 was needed as it would only repeat the line 74.